fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -2717,46 +2728,80 @@ ACTOR Future<Void> reply(CommitBatchContext* self) {
+	pContext->setupTraceBatch();
 
 	/////// Phase 1: Pre-resolution processing (CPU bound except waiting for a version # which is separately pipelined
 	/// and *should* be available by now (unless empty commit); ordered; currently atomic but could yield)
-	wait(CommitBatch::preresolutionProcessing(&context));
-	if (context.rejected) {
-		self->commitBatchesMemBytesCount -= currentBatchMemBytesCount;
+	pContext->stage = PRE_RESOLUTION;
+	wait(CommitBatch::preresolutionProcessing(pContext));
+	if (pContext->rejected) {
+		pContext->pProxyCommitData->commitBatchesMemBytesCount -= pContext->currentBatchMemBytesCount;
 		return Void();
 	}
 
 	/////// Phase 2: Resolution (waiting on the network; pipelined)
-	wait(CommitBatch::getResolution(&context));
+	pContext->stage = RESOLUTION;
+	wait(CommitBatch::getResolution(pContext));
 
 	////// Phase 3: Post-resolution processing (CPU bound except for very rare situations; ordered; currently atomic but
 	/// doesn't need to be)
-	wait(CommitBatch::postResolution(&context));
+	pContext->stage = POST_RESOLUTION;
+	wait(CommitBatch::postResolution(pContext));
 
 	/////// Phase 4: Logging (network bound; pipelined up to MAX_READ_TRANSACTION_LIFE_VERSIONS (limited by loop above))
-	wait(CommitBatch::transactionLogging(&context));
+	pContext->stage = TRANSACTION_LOGGING;
+	wait(CommitBatch::transactionLogging(pContext));
 
 	/////// Phase 5: Replies (CPU bound; no particular order required, though ordered execution would be best for
 	/// latency)
-	wait(CommitBatch::reply(&context));
+	pContext->stage = REPLY;
+	wait(CommitBatch::reply(pContext));
+
+	pContext->stage = COMPLETE;
+	return Void();
+}
+
+} // namespace CommitBatch
+
+ACTOR Future<Void> commitBatch(ProxyCommitData* pCommitData,
+                               std::vector<CommitTransactionRequest>* trs,
+                               int currentBatchMemBytesCount) {
+
+	state CommitBatch::CommitBatchContext context(pCommitData, trs, currentBatchMemBytesCount);
+
+	Future<Void> commit = CommitBatch::commitBatchImpl(&context);
+
+	// When encryption is enabled, cipher key fetching issue (e.g KMS outage) is detected by the
+	// encryption monitor. In that case, commit timeout is expected and timeout error is suppressed. But
+	// we still want to trigger recovery occasionally (with the COMMIT_PROXY_MAX_LIVENESS_TIMEOUT), in
+	// the hope that the cipher key fetching issue could be resolve by recovery (e.g, if one CP have
+	// networking issue connecting to EKP, and recovery may exclude the CP).
+	Future<Void> livenessTimeout = timeoutErrorIfCleared(
+	    commit, pCommitData->encryptionMonitor->degraded(), SERVER_KNOBS->COMMIT_PROXY_LIVENESS_TIMEOUT);
+
+	Future<Void> maxLivenessTimeout = timeoutError(livenessTimeout, SERVER_KNOBS->COMMIT_PROXY_MAX_LIVENESS_TIMEOUT);
+	try {
+		wait(maxLivenessTimeout);
+	} catch (Error& err) {
+		TraceEvent(SevInfo, "CommitBatchFailed").detail("Stage", context.stage).detail("ErrorCode", err.code());
+		throw;
```

Comment:
The original code uses `transformError` to throw `failed_to_progress()` error. We should preserve the same behavior here. BTW, "CommitBatchFailed" event doesn't tell which tlog failed or is slow.

Reply:
This is a serious problem (will cause recovery), so the level should be at least `SevWarnAlways`.